libs/s25main/world/GameWorld.cpp

```diff
@@ -413,43 +413,43 @@ void GameWorld::RecalcTerritory(const noBaseBuilding& building, TerritoryChangeR
             sizeChanges[oldOwner - 1]--;
     }
 
-    std::set<MapPoint, MapPointLess> ptsHandled;
+    const std::vector<MapPoint> ptsToHandled = GetAllNeighboursUnion(ptsWithChangedOwners);
+
     // Destroy everything from old player on all nodes where the owner has changed
-    for(const MapPoint& curMapPt : ptsWithChangedOwners)
+    for(const MapPoint& curMapPt : ptsToHandled)
     {
         // Destroy everything around this point as this is at best a border node where nothing should be around
         // Do not destroy the triggering building or its flag
-        // TODO: What about this point?
         const uint8_t owner = GetNode(curMapPt).owner;
-        for(const MapPoint neighbourPt : GetNeighbours(curMapPt))
-        {
-            if(ptsHandled.insert(neighbourPt).second)
-                DestroyPlayerRests(neighbourPt, owner, &building);
-        }
+        DestroyPlayerRests(curMapPt, owner, &building);
 
         if(gi)
             gi->GI_UpdateMinimap(curMapPt);
     }
 
     // Destroy remaining roads going through non-owned territory
-    for(const MapPoint& curMapPt : ptsWithChangedOwners)
+    for(const MapPoint& curMapPt : ptsToHandled)
     {
         // Skip if there is an object. We are looking only for roads going through, not ending here
         // (objects here are already destroyed and if the road ended there it would have been as well)
         if(GetNode(curMapPt).obj)
             continue;
         Direction dir;
         noFlag* flag = GetRoadFlag(curMapPt, dir);
-        if(!flag || flag->GetPlayer() + 1 == GetNode(curMapPt).owner)
+
+        if(!flag)
             continue;
+        if(flag->GetPlayer() + 1 == GetNode(curMapPt).owner && !IsBorderNode(curMapPt, GetNode(curMapPt).owner))
```

Comment:
Double use of potentially expensive `GetNode`, store:
```suggestion
        const uint8_t owner = GetNode(curMapPt).owner;
        if(flag->GetPlayer() + 1 == owner && !IsBorderNode(curMapPt, owner))
```